torch/fx/experimental/proxy_tensor.py

```diff
@@ -527,3 +529,58 @@ def wrap_fake_symbolic(x, sym_shape):
         return t
 
     return wrapped
+
+
+def get_torch_dispatch_modes():
+    modes = [torch._C._get_torch_dispatch_mode()]
+    if modes[-1] is None:
+        return list()
+    while modes[-1].inner is not None:
+        modes.append(modes[-1].inner)
+    return modes
+
+
+def get_isolated_graphmodule(func, args, kwargs):
+    """A helper function used to get the GraphModule for the given func.
+
+    It's expected to be used in the ProxyTensor tracing context.
+    It detaches the args and kwargs from the current tracer so that the trace of
+    the current graph module can be created without any side-effects.
+    """
+    # make_fx doesn't support kwargs, so we need to do this flattening
+    # and then unflatten the args before calling func
+    all_args, spec = pytree.tree_flatten((args, kwargs))
+
+    def wrapped(args):
+        fn_args, fn_kwargs = pytree.tree_unflatten(args, spec)
+        return func(*fn_args, **fn_kwargs)
+
+    unwrapped_all_args = [unwrap_elem(a) for a in all_args]
+
+    # Current implementation doesn't support the case when ProxyTensor is
+    # wrapped with another Tensor subclass
+    # See: https://github.com/pytorch/pytorch/pull/81764#issuecomment-1200472068
+    assert all(
+        getattr(a, "elem", None) is None
+        for a in unwrapped_all_args
+        if isinstance(a, torch.Tensor)
+    ), "ProxyTensor is wrapped with another Tensor subclass"
+
+
+    with contextlib.ExitStack() as stack:
+        modes = get_torch_dispatch_modes()
+        # Disable all torch dispatch modes
+        for mode in modes:
+            stack.enter_context(enable_torch_dispatch_mode(mode.inner, replace=mode))
+        assert torch._C._get_torch_dispatch_mode() is None
+
+        # Enable all torch dispatch modes except ProxyTorchDispatchMode
+        for mode in reversed([m for m in modes if not isinstance(m, ProxyTorchDispatchMode)]):
```

Comment:
@samdow, what do you think of the approach here modifying `.inner` and `.ancestors` to rebuild torch dispatch context skipping `ProxyTorchDispatchMode` instances?

Reply:
I think I mostly agree with @ezyang that I don't love this as an idea but I get that copying has some limitations that make this difficult (ideally we would want to copy while passing a new argument to the constructor...). So what about this idea:
(1) delete the inner and ancestor attributes for the copy of every mode in the stack (note: they cannot just be set to None but must be deleted in order to get the mode code to work)
(2) as we're doing this, one by one push the modes back onto the stack

This is still imperfect since we are altering the inner and ancestor elements of the stack but at least the mode mechanism is still the one in charge of the mode ordering instead of having this repeated code

Reply:
Also open to thoughts from either of you on this idea

Reply:
I am still advocating for just setting dynamically scoped variables on the context objects themselves. This saves you from (1) having to copy or (2) mutating the inner pointers resulting in weird behavior

Reply:
I'm happy with that. And to Ed's other point, I would be happy to have us restrict users updating inner but not sure if there's a clean way to